File: openex-platform/openex-api/src/Entity/Tag.php
<?php

namespace App\Entity;

use App\Entity\Base\BaseEntity;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;



/**
 * @ORM\Entity()
 * @ORM\Table(name="tags")
 */
class Tag extends BaseEntity
{

    /**
     * @ORM\Id
     * @ORM\Column(type="string")
     * @ORM\GeneratedValue(strategy="CUSTOM")
     * @ORM\CustomIdGenerator("doctrine.uuid_generator")
     */
    protected $tag_id;
    /**
     * @ORM\Column(type="string")
     */
    protected $tag_name;
    /**
     * @ORM\ManyToMany(targetEntity="Document", mappedBy="document_tags")
     * @var Documents[]
     */
    protected $tag_documents;

    public function __construct()
    {
        parent::__construct();
        $this->tag_documents = new ArrayCollection();
    }

    public function getTagId()
    {
        return $this->tag_id;
    }

    public function getTagName()
    {
        return $this->tag_name;
    }

    public function setTagName($name)
    {
        $this->tag_name = $name;
        return $this;
    }
}

File: openex-api/src/main/java/io/openex/database/model/Tag.java
package io.openex.database.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.openex.database.audit.ModelBaseListener;
import io.openex.helper.MultiModelDeserializer;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "tags")
@EntityListeners(ModelBaseListener.class)
public class Tag implements Base {
    @Id
    @Column(name = "tag_id")
    @GeneratedValue(generator = "UUID")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    @JsonProperty("tag_id")
    private String id;

    @Column(name = "tag_name")
    @JsonProperty("tag_name")
    private String name;

    @Column(name = "tag_color")
    @JsonProperty("tag_color")
    private String color;

    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(name = "documents_tags",
            joinColumns = @JoinColumn(name = "tag_id"),
            inverseJoinColumns = @JoinColumn(name = "document_id"))
    @JsonSerialize(using = MultiModelDeserializer.class)
    @JsonProperty("tags_documents")
    @Fetch(FetchMode.SUBSELECT)
    private List<Document> documents = new ArrayList<>();

    @JsonIgnore
    @Override
    public boolean isUserObserver(User user) {
        return true;
    }

    @Override
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name.toLowerCase();
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color.toLowerCase();
    }

    public List<Document> getDocuments() {
        return documents;
    }

    public void setDocuments(List<Document> documents) {
        this.documents = documents;
    }
}
